### PR TITLE
Observe layout guides

### LINKED ISSUE #6954: Compass underlaps top bar
In a standard view controller with a top bar and `automaticallyAdjustsScrollViewInsets` set to `YES`, the compass underlaps the top bar:

underlap

This is almost certainly a regression from iOS SDK v3.3.x. Relevant PRs from this release cycle: #6216 #6313 #6566.

/cc @boundsj @incanus

### PR DESCRIPTION
**Pull request by frederoni**
Fixes #6954

This PR adds observers to the top and bottom layout guides if necessary in order to re-layout after changes have occurred. (i.e due to showing/hiding a navigation bar)

It also replaces constraints with manual layout to avoid manipulating the view controller's view which is the developer's responsibility.

animation

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -1866,6 +1890,10 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
             }
         }
     }
+    else if ([keyPath isEqualToString:@"bounds"] && [object isKindOfClass:NSClassFromString(@"_UILayoutGuide")])
```
Comment (frederoni) [inline platform/ios/src/MGLMapView.mm]: Comparing against the private class _UILayoutGuide is not particularly future proof.

File: platform/ios/src/MGLMapView.mm [hunk 2]
```diff
@@ -122,6 +122,9 @@ typedef NS_ENUM(NSUInteger, MGLUserTrackingState) {
 
 const CGSize MGLAnnotationAccessibilityElementMinimumSize = CGSizeMake(10, 10);
 
+// Context for KVO observing UILayoutGuides.
+static char MGLLayoutGuidesUpdatedContext = 1;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This works, but the more common practice is to pass in a selector such as `_cmd`.

File: platform/ios/src/MGLMapView.mm [hunk 3]
```diff
@@ -655,6 +658,11 @@ - (void)dealloc
     [[NSNotificationCenter defaultCenter] removeObserver:self];
     [_attributionButton removeObserver:self forKeyPath:@"hidden"];
 
+    if (_isObservingLayoutGuides) {
+        [(UIView *)self.viewControllerForLayoutGuides.topLayoutGuide removeObserver:self forKeyPath:@"bounds" context:(void*)&MGLLayoutGuidesUpdatedContext];
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Is topLayoutGuide really a UIView? I'd expect it to be a UILayoutGuide.

File: platform/ios/src/MGLMapView.mm (hunk 2 above)
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Er, I may be mistaken about `_cmd`. But in any case, the important thing is to ensure that the context’s _value_ is unique. If we later add a key-value observer with a different constant set to 1, suddenly there’s an opportunity for crosstalk.

A simple fix is to [set the context equal to its own address](http://nshipster.com/key-value-observing/#correct-context-declarations).

**Comment by 1ec5** [comment]
Postponing to v3.4.1, per https://github.com/mapbox/mapbox-gl-native/issues/6954#issuecomment-273283732.

**Comment by 1ec5** [comment]
Since this PR is scheduled for iOS SDK v3.5.0, it needs to be rebased and retargeted at the release-ios-v3.5.0-android-v5.0.0 branch.

**Comment by frederoni** [comment]
- Rebased and retargeted against the 3.6 release branch.
- Fixed layout for the recently added scale bar.

observe

@boundsj @fabian-guerra 👀

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -931,6 +856,38 @@ - (void)layoutSubviews
     [self updateUserLocationAnnotationView];
 }
 
+- (void)layoutOrnaments
+{
+    CGFloat x, y;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This is OK, but just a stylistic tip: `x` and `y` variables that get repurposed multiple times carry the risk that we might fail to update one of these variables and its previous value may “bleed” into a subsequent use of it.

Instead, I sometimes take advantage of the struct initializer syntax to keep the values scoped to where they’re used. For example:

```objc
self.compassView.center = {
    CGRectGetWidth(self.bounds) - CGRectGetMidX(self.compassView.bounds) - self.contentInset.right - 5,
    CGRectGetMidY(self.compassView.bounds) + self.contentInset.top + 5,
};
```

If you’re concerned about accidentally swapping `x` and `y`, you can also use the C99 designated initializer syntax:

```objc
self.compassView.center = {
    .x = CGRectGetWidth(self.bounds) - CGRectGetMidX(self.compassView.bounds) - self.contentInset.right - 5,
    .y = CGRectGetMidY(self.compassView.bounds) + self.contentInset.top + 5,
};
```

**Pull request merged**